taichi/backends/opengl/opengl_api.h

```diff
@@ -11,17 +11,21 @@ TLANG_NAMESPACE_BEGIN
 
 namespace opengl {
 
-struct GLProgram;
 void initialize_opengl();
 bool is_opengl_api_available();
 void create_glsl_root_buffer(size_t size);
 void begin_glsl_kernels(const std::vector<IOV> &iov);
-void launch_glsl_kernel(GLProgram *program, int num_groups);
 void end_glsl_kernels(const std::vector<IOV> &iov);
-GLProgram *compile_glsl_program(std::string source);
 int opengl_get_threads_per_group();
 extern bool opengl_has_GL_NV_shader_atomic_float;
 
+struct GLProgram;
+struct GLProgramIFace {
+  std::shared_ptr<GLProgram> glsl;
```

Comment:
It seems to me that `GLProgramIFace` holds the ownership of `glsl`. Maybe we should consider using a `unique_ptr` here instead of `shared_ptr`?

Reply:
We can use `unique_ptr` if `CompiledKernel` is not copied...the compiler complains me about copy constructors.

Reply:
Let me see... do you mean somewhere around https://github.com/taichi-dev/taichi/blob/949ef7f961e1ed23836e01200fa26f43ae9463f0/taichi/codegen/codegen_opengl.cpp#L278?

Reply:
Sorry, but could you provide a permalink to the place where the copy happens?

Reply:
Oh, sorry! I'm taking line ref from another PR... It's actually in:
https://github.com/taichi-dev/taichi/blob/949ef7f961e1ed23836e01200fa26f43ae9463f0/taichi/codegen/codegen_opengl.cpp#L695-L697

Reply:
Is C++ expecting me to use the `struct KernelGen : public CompiledProgram` attempt again?

Reply:
How to save the `compiled_program_` pointer while safely deleting the whole `KernelGen`?

Reply:
I see what's happening. What you really need is not a copy here, but a move. In my understanding after `get_compiled_program`, `KernelGen` no longer needs `compiled_program_` any more. What we can do here is to transfer the ownership of `compiled_program_` from `KernelGen` to the invoker. 

The easiest way to do this now, is to use `std::unique_ptr<CompiledProgram> compiled_program_` in `KernelGen` so that you can do something like 
```C++
std::unique_ptr<CompiledProgram> get_compiled_program() const { 
   return std::move(compiled_program_);
} 
```

Reply:
To be honest, `shared_ptr` have no disadvantages (1->2->1), I would like to use it, instead of making all things hard, `unique_ptr` is aim to help devs (1->1->1), not to make us headache (1->0->1).

Reply:
...and then no copy is needed for `CompiledGLSL`, so you can use `std::unique_ptr<GLProgram>`.

Reply:
> If `shared_ptr` have no disadvantages (1->2->1), I would like to use it, instead of making all things hard, `unique_ptr` is aim to help devs (1->1->1), not to make us headache (1->0->1).

The disadvantages of `std::shared_ptr` are
1) It is much slower than `std::unique_ptr` since it has to be thread-safe;
2) It makes ownership of your members unclear...

In smaller projects, these issues might not be very pronounced, but as Taichi is growing bigger we should pay attention to these things (especially 2).